javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/InstrumentationModuleClassLoader.java

```diff
@@ -0,0 +1,232 @@
+          } else {
+            result =
+                AccessController.doPrivileged(
+                    (PrivilegedAction<Class<?>>) () -> defineClassWithPackage(name, bytecode));
+          }
+        }
+      }
+      if (result == null) {
+        result = tryLoad(agentOrExtensionCl, name);
+      }
+      if (result == null) {
+        result = tryLoad(instrumentedCl, name);
+      }
+
+      if (result != null) {
+        if (resolve) {
+          resolveClass(result);
+        }
+        return result;
+      } else {
+        throw new ClassNotFoundException(name);
+      }
+    }
+  }
+
+  private static Class<?> tryLoad(ClassLoader cl, String name) {
+    try {
+      return cl.loadClass(name);
+    } catch (ClassNotFoundException e) {
+      return null;
+    }
+  }
+
+  @Override
+  public URL getResource(String resourceName) {
+    String className = resourceToClassName(resourceName);
+    if (className != null) {
+      // for classes use the same precedence as in loadClass
+      ClassCopySource injected = getInjectedClass(className);
+      if (injected != null) {
+        return injected.getUrl();
+      }
+      URL fromAgentCl = agentOrExtensionCl.getResource(resourceName);
+      if (fromAgentCl != null) {
+        return fromAgentCl;
+      }
+      return instrumentedCl.getResource(resourceName);
+    } else {
+      // delegate to just the default parent (the agent classloader)
+      return super.getResource(resourceName);
+    }
+  }
+
+  @Override
+  public Enumeration<URL> getResources(String resourceName) throws IOException {
+    String className = resourceToClassName(resourceName);
+    if (className != null) {
+      URL resource = getResource(resourceName);
+      List<URL> result =
+          resource != null ? Collections.singletonList(resource) : Collections.emptyList();
+      return Collections.enumeration(result);
+    }
+    return super.getResources(resourceName);
+  }
+
+  @Nullable
+  private static String resourceToClassName(String resourceName) {
+    if (!resourceName.endsWith(".class")) {
+      return null;
+    }
+    String className = resourceName;
+    if (className.startsWith("/")) {
+      className = className.substring(1);
+    }
+    className = className.replace('/', '.');
+    className = className.substring(0, className.length() - ".class".length());
+    return className;
+  }
+
+  @Nullable
+  private ClassCopySource getInjectedClass(String name) {
+    ClassCopySource alwaysInjected = ALWAYS_INJECTED_CLASSES.get(name);
+    if (alwaysInjected != null) {
+      return alwaysInjected;
+    }
+    return additionalInjectedClasses.get(name);
+  }
+
+  private Class<?> defineClassWithPackage(String name, byte[] bytecode) {
+    int lastDotIndex = name.lastIndexOf('.');
+    if (lastDotIndex != -1) {
+      String packageName = name.substring(0, lastDotIndex);
+      if (findPackage(packageName) == null) {
+        definePackage(packageName, null, null, null, null, null, null, null);
```

Comment:
I believe that on jdk8 this might not be safe and may result in `IllegalArgumentException` from https://github.com/openjdk/jdk8u/blob/587090ddc17c073d56f4d3f52b61f6477d6322b0/jdk/src/share/classes/java/lang/ClassLoader.java#L1587. `getPackage` searches for package from the parent loader and then from the current loader. As this is a child first loader when it starts to define class parent loader is not locked and can define the same package at the same time which would result in a race where package in parent loader is not define when `findPackage` is called but already is defined when `definePackage` is called. For this race to happen this class loader would need to define a class that is in the same package as a class defined in the parent loader at the same time, IDK whether this is really possible. If this were a parallel capable class loader then there could be more ways it could hit a race here and get the same `IllegalArgumentException`.

Reply:
Any idea how to guard against this?
I can only think of wrapping the `definePackage` in a `try ... catch(IllegalArgumentException)` and then printing the occurring exception to the debug logs.

In the elastic-agent we use a bytebuddy `ByteArrayInjectingClassloader`. This comes with the downside that (a) all the bytecode of injected classes needs to be loaded eagerly and (b) the resource lookup for `.class` files is either inconsistent or costs heap because the bytecode `byte[]` need to be kept in memory.
This is something I wanted to improve here by using the `ClassCopySource` instead. I did my best of basically "inlining" the required logic from `ByteArrayClassloader` for defining classes here.

You can find the bytebuddy code for defining the package [here](https://github.com/raphw/byte-buddy/blob/5e3f7fcd47ec135d3e9cf6eeb5c0135a8b91dcc4/byte-buddy-dep/src/main/java/net/bytebuddy/dynamic/loading/ByteArrayClassLoader.java#L673) it is in theory prone to the same problem, but we have never encountered it in the elastic apm agent so far.

Reply:
tomcat ignore is https://github.com/apache/tomcat/blob/39c388b51817d346031189a8a942340b6e46b940/java/org/apache/catalina/loader/WebappClassLoaderBase.java#L2280
URLClassLoader catches it and checks whether package is defined before failing https://github.com/openjdk/jdk8u/blob/587090ddc17c073d56f4d3f52b61f6477d6322b0/jdk/src/share/classes/java/net/URLClassLoader.java#L437
I think either of these approaches is fine. In my opinion logging the exception isn't necessary, if you feel it is necessary do make sure to log it at a level where it doesn't get printed to stdout unless debug is enabled.